### PR TITLE
Feature nice paper review form

### LINKED ISSUE #27: Dynamic paper selection form

### LINKED ISSUE #28: Scrapers for automatically selecting papers

### PR DESCRIPTION
**Pull request by WNoort**
Fixes #27 
Fixes #28

### CODE REVIEW

File: openreview/apps/main/forms.py [hunk 1]
```diff
@@ -63,29 +71,33 @@ def save(self, commit=True, **kwargs):
...
-            for keyword in self.cleaned_data["keywords"]:
-                if keyword.id is None:
-                    keyword.save()
-            paper.keywords.add(*self.cleaned_data["keywords"])
-
-        return paper
+        try:
+            existing_paper = Paper.objects.get(doc_id=self.cleaned_data.get("doc_id"))
+            return existing_paper
+        except ObjectDoesNotExist:
```
Comment (martijnbastiaan) [inline openreview/apps/main/forms.py]: existing_paper --> inline

Paper.DoesNotExist is je exception

File: openreview/apps/main/tests/reviewform.py [hunk 2]
```diff
@@ -103,8 +158,9 @@ def test_paper_gets_committed(self):
         self.wd.wait_for_css("#id_title")
 
         # Select the right paper
+        self.wd.find_css("#id_type option[value=\"manually\"]").click()
```
Comment (martijnbastiaan) [inline openreview/apps/main/tests/reviewform.py]:
```
a = "een 'string'"
b = 'een "string"'
```

File: openreview/apps/main/views.py [hunk 3]
```diff
@@ -22,9 +24,19 @@ def frontpage(request):
 @login_required
 def add_review(request):
     data = request.POST if "add_review" in request.POST else None
-    paper_form = PaperForm(data=data)
+
     review_form = ReviewForm(data=data, user=request.user)
 
+    if data and data.get('type') == 'arxiv':
+        #TODO potentially unsave?
+        scraper = (ArXivScraper().parse(data.get('doc_id'))).get_results()
```
Comment (martijnbastiaan) [inline openreview/apps/main/views.py]: Buitenste haakjes rond (ArXivScraper.....) kunnen weg

File: openreview/apps/papers/scrapers.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    def parse(self, doc_id):
+        if self.caching and cache.get("arxiv-res-{id}".format(id=doc_id)):
+            self.fields = cache.get("arxiv-res-{id}".format(id=doc_id))
+        else:
+            self.fields['doc_id'] = doc_id
+            self.fields['urls'] = "http://arxiv.org/abs/{id}".format(id=doc_id)
+            self.bs = BeautifulSoup(PaperMetaScraper.urlopen(self.fields['urls']))
+            self.fields['title'] = self.bs.select("div.leftcolumn h1.title")[0].span.nextSibling
+            #I need a functional language
+            self.fields['authors'] = list(map(lambda x: x.text, self.bs.select("div.leftcolumn div.authors a")))
```
Comment (martijnbastiaan) [inline openreview/apps/papers/scrapers.py]: self.fields['authors'] = [x.text for x in self.bs.select('.......')]

File: openreview/apps/papers/views.py [hunk 5]
```diff
@@ -171,3 +175,17 @@ def delete(self, request, **kwargs):
         return HttpResponse("OK", status=200)
 
 
+@cache_page(60*10)
+def doi_scraper(request, id):
+    return HttpResponse(json.JSONEncoder().encode({"error": "Invalid document identifier"}),
```
Comment (martijnbastiaan) [inline openreview/apps/papers/views.py]: json.dumps()

**Comment by martijnbastiaan** [comment]
- defaultdict is handig in veel gevallen (nu niet helemaal toepasbaar), maar: http://docs.python.org/2/library/collections.html#collections.defaultdict
- model Paper gebruiken in PaperMetaScraper
- spatie na comment

**Comment by martijnbastiaan** [comment]
Oh, en de javascript: verwijder de potentiele XSS aanvalsvectoren. Over de stijl van de js-code heb ik verder wel een mening, maar geen sterke dus dat laten we lekker hangen :-).

Zie:
1. https://github.com/open-review/open-review/blob/master/openreview/apps/papers/templates/papers/comment.html#L14
2. https://github.com/open-review/open-review/blob/master/openreview/apps/papers/static/papers/writing.coffee#L26

**Comment by WNoort** [comment]
Het is aardig gelukt, behalve het model Paper gebruiken in de scrapers. Als ik namelijk authors wil toevoegen, moet ik een ManyToMany relation gebruiken en die wil dat ik de objecten al in de DB heb staan. Dat lijkt me nogal ongewenst.

Verder lijken de tests goed, maar ik kreeg wel opeens een keer een stackoverflow exception:

Fatal Python error: Cannot recover from stack overflow.

Current thread 0x00007f8209d66740:
  File "/usr/local/lib/python3.3/dist-packages/bs4/element.py", line 664 in **getnewargs**
  File "/usr/local/lib/python3.3/dist-packages/memcache.py", line 760 in _val_to_store_info
  File "/usr/local/lib/python3.3/dist-packages/memcache.py", line 806 in _unsafe_set
  File "/usr/local/lib/python3.3/dist-packages/memcache.py", line 825 in _set
  File "/usr/local/lib/python3.3/dist-packages/memcache.py", line 573 in set
  File "/usr/local/lib/python3.3/dist-packages/django/core/cache/backends/memcached.py", line 82 in set
  File "/home/willem/code/open-review/openreview/apps/papers/scrapers.py", line 45 in parse
  File "/home/willem/code/open-review/openreview/apps/main/views.py", line 32 in add_review
  File "/usr/local/lib/python3.3/dist-packages/django/contrib/auth/decorators.py", line 22 in _wrapped_view
  File "/usr/local/lib/python3.3/dist-packages/django/core/handlers/base.py", line 114 in get_response
  File "/usr/local/lib/python3.3/dist-packages/django/test/client.py", line 109 in __call__
  File "/usr/local/lib/python3.3/dist-packages/django/test/client.py", line 426 in request
  File "/usr/local/lib/python3.3/dist-packages/django/test/client.py", line 302 in post
  File "/usr/local/lib/python3.3/dist-packages/django/test/client.py", line 483 in post
  File "/home/willem/code/open-review/openreview/apps/main/tests/reviewform.py", line 116 in test_form_filled_in_automatically
  File "/usr/lib/python3.3/unittest/case.py", line 384 in _executeTestPart
  File "/usr/lib/python3.3/unittest/case.py", line 439 in run
  File "/usr/lib/python3.3/unittest/case.py", line 491 in __call__
  File "/usr/lib/python3.3/unittest/suite.py", line 105 in run
  File "/usr/lib/python3.3/unittest/suite.py", line 67 in **call**
  File "/usr/lib/python3.3/unittest/runner.py", line 168 in run
  File "/usr/local/lib/python3.3/dist-packages/django/test/runner.py", line 113 in run_suite
  File "/usr/local/lib/python3.3/dist-packages/django/test/runner.py", line 146 in run_tests
  File "/usr/local/lib/python3.3/dist-packages/django/core/management/commands/test.py", line 88 in handle
  File "/usr/local/lib/python3.3/dist-packages/south/management/commands/test.py", line 8 in handle
  File "/usr/local/lib/python3.3/dist-packages/django/core/management/base.py", line 285 in execute
  File "/usr/local/lib/python3.3/dist-packages/django/core/management/commands/test.py", line 71 in execute
  File "/usr/local/lib/python3.3/dist-packages/django/core/management/base.py", line 242 in run_from_argv
  File "/usr/local/lib/python3.3/dist-packages/django/core/management/commands/test.py", line 50 in run_from_argv
  File "/usr/local/lib/python3.3/dist-packages/django/core/management/**init**.py", line 392 in execute
  File "/usr/local/lib/python3.3/dist-packages/django/core/management/**init**.py", line 399 in execute_from_command_line
  File "./manage.py", line 10 in <module>
Afgebroken

**Comment by martijnbastiaan** [comment]
Je zou een 'namedtuple' kunnen teruggeven in je scraper code. De code eromheen kan dan vervolgens besluiten het in de database te zetten. Vb:

``` python
ScrapedPaper = namedtuple("ScrapedPaper", ['paper', 'authors', 'keywords'])
```

Kan de stack overflow misschien komen door beautifulsoup? Ik heb al vaker gehad dat een document met relatief weinig 'diepte' BS tot een stackoverflow kunnen drijven.

**Comment by WNoort** [comment]
Die namedtuple snap ik niet en de stack overflow kan prima komen door BS

**Comment by martijnbastiaan** [comment]
Woeps, ik had niet door dat dit nog niet helemaal af was. Heb je nog hulp nodig? Kan ik wat overnemen?

**Comment by WNoort** [comment]
Het werkt gewoon. Wat mij betreft is het goed

**Comment by martijnbastiaan** [comment]
Helemaal goed, dank!

**Pull request merged**